src/runtime/msan.cpp

```diff
@@ -0,0 +1,93 @@
+#include "HalideRuntime.h"
+
+extern "C" {
+
+namespace Halide { namespace Runtime { namespace Internal {
+
+struct Dimension {
+  ptrdiff_t stride_bytes;
+  size_t extent;
+};
+
+WEAK void halide_msan_annotate_buffer_is_initialized_internal(
+    void *user_context, const Dimension *dims, int dims_count, int level, uint8_t* host) {
+
+    const ptrdiff_t stride_bytes = dims[level].stride_bytes;
+    const size_t extent = dims[level].extent;
+    if (level == dims_count - 1) {
+        const size_t swath_bytes = stride_bytes * extent;
+        halide_msan_annotate_memory_is_initialized(user_context, host, swath_bytes);
+        return;
+    }
+
+    for (size_t i = 0; i < extent; ++i) {
+        halide_msan_annotate_buffer_is_initialized_internal(
+            user_context, dims, dims_count, level + 1, host + i * stride_bytes);
+    }
+}
+
+}}} // namespace Halide::Runtime::Internal
+
+WEAK void halide_msan_annotate_memory_is_initialized(void *user_context, void *ptr, size_t len) {
+    // Default implementation does nothing; you must override this and supply (e.g.)
+    //    ANNOTATE_MEMORY_IS_INITIALIZED(ptr, len);
+    // to mark the memory as initialized.
+}
+
+// Default implementation marks the data pointed to by the buffer_t as initialized
+// (but *not* the buffer_t itself); it takes pains to only mark the active memory ranges
+// (skipping padding), and sorting into ranges to always mark the smallest number of
+// ranges, in monotonically increasing memory order.
+WEAK void halide_msan_annotate_buffer_is_initialized(void *user_context, void *v) {
+    halide_print(user_context, "halide_msan_annotate_buffer_is_initialized\n");
+    buffer_t *b = (buffer_t *)v;
+    if (b == NULL) {
+        return;
+    }
+
+    enum { kMaxDims = 4 };
+    Dimension dims[kMaxDims + 1];
+    int dim_count = 0;
+    for (int i = 0; i < kMaxDims; ++i) {
+        if (b->extent[i] == 0) {
+          break;
+        }
+        dims[dim_count].stride_bytes = static_cast<ptrdiff_t>(b->stride[i]) * b->elem_size;
+        dims[dim_count].extent = b->extent[i];
+        dim_count++;
+    }
+    if (dim_count == 0) {
+        return;
+    }
+
+    // Sort in decreasing order of stride_bytes.
```

Comment:
Should we bother with all of this cleverness? It makes things more efficient... but it's msan. It's going to be slow anyways, and it's more important to be correct for this use. Consider that this code will need to be modified fairly significantly for the arbitrary dimensions PR. I think maybe we should just collapse dense dimensions (i.e. one annotate call for all of stride 1 dimensions) and otherwise do the naive thing.